windows/src/engine/keyman32/appint/appint.cpp

```diff
@@ -135,16 +137,47 @@ void AppContext::CopyFrom(AppContext *source)   // I3575
 
 void AppContext::Set(const WCHAR *buf)
 {
-	const WCHAR *p;
-	WCHAR *q;
-	for(p = buf, q = CurContext; *p && (INT_PTR)(q-CurContext) < MAXCONTEXT - 1; p++, q++)
-	{
-		*q = *p;
-		if(*p >= 0xD800 && *p <= 0xDBFF) { *(++q) = *(++p); }
-	}
-	*q = 0;
-  pos = (int)(INT_PTR)(q-CurContext);  // I1129 - Irregular behaviour with context rules
-	CurContext[MAXCONTEXT-1] = 0;
+  const WCHAR *p;
+  WCHAR *q;
+
+  // We may be past a buffer longer than our internal
+  // buffer. So we shift to make sure we capture the end
+  // of the string, not the start
+  p = wcschr(buf, 0);
+  q = (WCHAR *)p;
+  while (p != NULL && p > buf && (intptr_t)(q - p) < MAXCONTEXT - 1) {
+    p = decxstr((WCHAR *)p, (WCHAR *)buf);
+  }
+
+  // If the first character in the buffer is a surrogate pair,
+  // or a deadkey, our buffer may be too long, so move to the
+  // next character in the buffer
+  if ((intptr_t)(q - p) > MAXCONTEXT - 1) {
+    p = incxstr((WCHAR *)p);
+  }
+
+  for (q = CurContext; *p; p++, q++) {
+    *q = *p;
+  }
+
+  *q = 0;
+  pos = (int)(intptr_t)(q - CurContext);
+  CurContext[MAXCONTEXT - 1] = 0;
+
+  //////
+  //const WCHAR *p;
+  //WCHAR *q;
+  //for (p = buf, q = CurContext; *p && (INT_PTR)(q - CurContext) < MAXCONTEXT - 1; p++, q++) {
+  //  *q = *p;
+  //  if (*p >= 0xD800 && *p <= 0xDBFF) {
+  //    *(++q) = *(++p);
+  //  }
+  //}
+  //*q                         = 0;
+  //pos                        = (int)(INT_PTR)(q - CurContext);  // I1129 - Irregular behaviour with context rules
+  //CurContext[MAXCONTEXT - 1] = 0;
```

Comment:
Please remove commented code

Reply:
Oh no that is a worry I know I removed that but must not off hit Ctrl+S before pushing it :(.